casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala

```diff
@@ -192,24 +192,18 @@ object Proposer {
                          (future.pure ||^ expired.pure ||^ replayAttack).not
                      }
                      .map(_.map(_._1))
-        deploys <- {
-          val dummy = dummyDeployOpt
-            .traverse {
-              case (privateKey, term) =>
-                val deployData = ConstructDeploy.sourceDeployNow(
-                  source = term,
-                  sec = privateKey,
-                  vabn = nextBlockNum - 1,
-                  shardId = shardId
-                )
-                BlockDagStorage[F].addDeploy(deployData).as(List(deployData.sig))
-            }
-          OptionT
-            .fromOption(pooledOk.nonEmpty.guard[Option].as(pooledOk))
-            .orElseF(dummy)
-            .value
-            .map(_.getOrElse(List()))
-        }
+        deploys <- for {
+                    dummy <- dummyDeployOpt.traverse {
+                              case (privateKey, term) =>
+                                val deployData = ConstructDeploy.sourceDeployNow(
+                                  source = term,
+                                  sec = privateKey,
+                                  vabn = nextBlockNum - 1,
+                                  shardId = shardId
+                                )
+                                BlockDagStorage[F].addDeploy(deployData).as(List(deployData.sig))
+                            }
+                  } yield Option(pooledOk).filter(_.nonEmpty).orElse(dummy).getOrElse(List.empty)
```

Comment:
This will add dummy deploy into deploy storage even if it is not required.
Maybe you can use OptionT instead of Option and declare `dummy` as a suspended value (as in my initial code).